crates/matrix-sdk-crypto/src/backups/mod.rs

```diff
@@ -555,6 +555,58 @@ impl BackupMachine {
 
         (backup, session_record)
     }
+
+    /// Import the given room keys into our store.
+    ///
+    /// # Arguments
+    ///
+    /// * `room_keys` - A list of previously exported keys that should be
+    /// imported into our store. If we already have a better version of a key
+    /// the key will *not* be imported.
+    ///
+    /// Returns a [`RoomKeyImportResult`] containing information about room keys
+    /// which were imported.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # use std::io::Cursor;
+    /// # use matrix_sdk_crypto::{OlmMachine, decrypt_room_key_export};
+    /// # use ruma::{device_id, user_id};
+    /// # let alice = user_id!("@alice:example.org");
+    /// # async {
+    /// # let machine = OlmMachine::new(&alice, device_id!("DEVICEID")).await;
+    /// # let export = Cursor::new("".to_owned());
+    /// let exported_keys = decrypt_room_key_export(export, "1234")?;
+    /// machine.backup_machine().import_backed_up_room_keys(exported_keys, |_, _| {}).await?;
+    /// # anyhow::Ok(())
+    /// # }
+    /// ```
+    pub async fn import_backed_up_room_keys(
+        &self,
+        room_keys: BTreeMap<OwnedRoomId, BTreeMap<String, BackedUpRoomKey>>,
+        progress_listener: impl Fn(usize, usize),
+    ) -> Result<RoomKeyImportResult, CryptoStoreError> {
+        let mut decrypted_room_keys = vec![];
+
+        for (room_id, room_keys) in room_keys {
+            for (session_id, room_key) in room_keys {
+                let room_key = ExportedRoomKey {
+                    algorithm: room_key.algorithm,
+                    room_id: room_id.to_owned(),
+                    sender_key: room_key.sender_key,
+                    session_id,
+                    session_key: room_key.session_key,
+                    sender_claimed_keys: room_key.sender_claimed_keys,
+                    forwarding_curve25519_key_chain: room_key.forwarding_curve25519_key_chain,
+                };
+
+                decrypted_room_keys.push(room_key);
```

Comment:
I suppose there can be lots of room keys, given a large room with lots of activity. Would it make sense to guesstimate the size of the `decrypted_room_keys` vector ahead of time, to avoid reallocations when running this `push` statement?

Reply:
That's a good idea, the API to download room keys has no pagination support. You can download all room keys, all room keys for a certain room, or one room key.

The method is almost guaranteed to be called with way too many room keys :sweat_smile: